src/xds/provider/agones.rs

```diff
@@ -0,0 +1,438 @@
+use prost::Message;
+use prost_types::{Any, Struct};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap};
+
+use crate::xds::{
+    config::{
+        cluster::v3::{
+            cluster::{ClusterDiscoveryType, DiscoveryType},
+            Cluster,
+        },
+        core::v3::{
+            address,
+            socket_address::{PortSpecifier, Protocol as SocketProtocol},
+            Address, Metadata, SocketAddress,
+        },
+        endpoint::v3::{
+            lb_endpoint::HostIdentifier, ClusterLoadAssignment, Endpoint, LbEndpoint,
+            LocalityLbEndpoints,
+        },
+        listener::v3::{FilterChain, Listener},
+    },
+    service::discovery::v3::DiscoveryResponse,
+    ResourceType,
+};
+
+const QUILKIN_CONFIGMAP: &str = "quilkin-config";
+const DEFAULT_CLUSTER_NAME: &str = "default-quilkin-cluster";
+const QUILKIN_URL: &str = "quilkin.dev";
+const QUILKIN_TOKEN_KEY: &str = "tokens";
+const QUILKIN_TOKEN_LABEL: &str = "quilkin.dev/tokens";
+
+pub struct AgonesProvider {
+    config: kube::Api<k8s_openapi::api::core::v1::ConfigMap>,
+    gameservers: kube::Api<GameServer>,
```

Comment:
Rather than using `kube::Api`, we should use [Reflectors](https://github.com/kube-rs/kube-rs#reflectors) instead, such that we have a local cache that is updated on changes, that we can poll for free with no performance hit to the K8s API control plane (In client-go we use Informers and Listeners, but the concept is the same).

Constant polling against the K8s API server is unfortunately a great way to make the API control plane (and etcd) really upset and will cause the performance of the entire cluster to plummet.